CRM/Contribute/BAO/Contribution.php

```diff
@@ -5471,95 +5472,82 @@ protected static function isPaymentInstrumentChange(&$params, $pendingStatuses)
    * @throws \CiviCRM_API3_Exception
    */
   protected static function updateMembershipBasedOnCompletionOfContribution($contribution, $primaryContributionID, $changeDate) {
-    $contribution->loadRelatedMembershipObjects();
-    if (empty($contribution->_relatedObjects['membership'])) {
-      return;
+    // Load memberships from contribution info. There may be multiple memberships (or none)
+    $membershipFields = array('id', 'contact_id', 'is_test', 'membership_type_id', 'contribution_recur_id', 'status_id.is_current_member',
+      'start_date', 'join_date', 'end_date', 'status_id');
+    $memberships = array();
+    if (!empty($contribution->contribution_recur_id)) {
+      // Load memberships associated with recurring contribution
+      $membershipResult = civicrm_api3('Membership', 'get', array(
+        'contribution_recur_id' => $contribution->contribution_recur_id,
+        'return' => $membershipFields,
+        'options' => ['limit' => 0],
+      ));
     }
-    $memberships = $contribution->_relatedObjects['membership'];
```

Comment:
@mattwire if we could find a way to do a PR that just changes the way the memberships are loaded (as an array via the api rather than the loadRelatedMemberships I feel like we could get that reviewed & simplify this down to something reviewable

Reply:
@eileenmcnaughton The problem with that is that the whole function (updateMembershipBasedOnCompletionOfContribution) was a complete mess and this PR was meant to be an extraction on the way to #11552 which cleans up and resolves various issues with this part of the code.  By locking in the actual functionality with a unit test (and it being used to fix real bugs on sites) I was hoping to get this merged as a step on the way.  Given the number of orgs that use memberships (and report issues around auto-renewal) I had hoped it would get a little more attention from some of those orgs.

This PR was attempting to do exactly what you request, namely change the way memberships are loaded from the `loadRelatedMemberships()` to using the API, but the whole function is tangled up with use of the "objects" returned by `loadRelatedMemberships()` that you'd end up retrieving via the array, then converting to objects and making it even more tangled up!